File: src/shared/types/monitoring.ts
export interface SystemMetrics {
  distro_name: string;
  timestamp: string;
  cpu: CpuMetrics;
  memory: MemoryMetrics;
  disk: DiskMetrics;
  network: NetworkMetrics;
  context_switches?: number;
  disk_io?: DiskIoMetrics;
  tcp_connections?: TcpConnectionMetrics;
  gpu?: GpuMetrics;
}

export interface DiskIoMetrics {
  read_bytes_per_sec: number;
  write_bytes_per_sec: number;
}

export interface TcpConnectionMetrics {
  established: number;
  time_wait: number;
  listen: number;
}

export interface GpuMetrics {
  utilization_percent: number | null;
  vram_used_bytes: number | null;
  vram_total_bytes: number | null;
}

export interface CpuMetrics {
  usage_percent: number;
  per_core: number[];
  load_average: [number, number, number];
}

export interface MemoryMetrics {
  total_bytes: number;
  used_bytes: number;
  available_bytes: number;
  cached_bytes: number;
  swap_total_bytes: number;
  swap_used_bytes: number;
}

export interface DiskMetrics {
  total_bytes: number;
  used_bytes: number;
  available_bytes: number;
  usage_percent: number;
}

export interface NetworkMetrics {
  interfaces: InterfaceStats[];
}

export interface InterfaceStats {
  name: string;
  rx_bytes: number;
  tx_bytes: number;
  rx_packets: number;
  tx_packets: number;
}

// --- Historical metrics types ---

export type TimeRange = "live" | "1h" | "6h" | "24h";

export interface MetricsHistoryPoint {
  timestamp: string;
  cpu_avg: number;
  cpu_min?: number;
  cpu_max?: number;
  mem_used_bytes: number;
  mem_total_bytes: number;
  disk_usage_percent: number;
  net_rx_rate: number;
  net_tx_rate: number;
  swap_used_bytes?: number;
  swap_total_bytes?: number;
  context_switches?: number;
  disk_io_read_bytes?: number;
  disk_io_write_bytes?: number;
  tcp_established?: number;
  tcp_time_wait?: number;
  tcp_listen?: number;
  gpu_utilization?: number;
  gpu_vram_used?: number;
  gpu_vram_total?: number;
}

export interface MetricsHistoryResponse {
  distro_name: string;
  granularity: "raw" | "1m";
  points: MetricsHistoryPoint[];
}

// --- Alert types ---

export type AlertType = "cpu" | "memory" | "disk";

export interface AlertThreshold {
  alert_type: AlertType;
  threshold_percent: number;
  enabled: boolean;
}

export interface AlertRecord {
  id: number;
  distro_name: string;
  alert_type: AlertType;
  threshold: number;
  actual_value: number;
  timestamp: string;
  acknowledged: boolean;
}

File: src-tauri/src/domain/entities/monitoring.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SystemMetrics {
    pub distro_name: String,
    pub timestamp: DateTime<Utc>,
    pub cpu: CpuMetrics,
    pub memory: MemoryMetrics,
    pub disk: DiskMetrics,
    pub network: NetworkMetrics,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_switches: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub disk_io: Option<DiskIoMetrics>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tcp_connections: Option<TcpConnectionMetrics>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gpu: Option<GpuMetrics>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CpuMetrics {
    pub usage_percent: f64,
    pub per_core: Vec<f64>,
    pub load_average: [f64; 3],
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryMetrics {
    pub total_bytes: u64,
    pub used_bytes: u64,
    pub available_bytes: u64,
    pub cached_bytes: u64,
    pub swap_total_bytes: u64,
    pub swap_used_bytes: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiskMetrics {
    pub total_bytes: u64,
    pub used_bytes: u64,
    pub available_bytes: u64,
    pub usage_percent: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetworkMetrics {
    pub interfaces: Vec<InterfaceStats>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InterfaceStats {
    pub name: String,
    pub rx_bytes: u64,
    pub tx_bytes: u64,
    pub rx_packets: u64,
    pub tx_packets: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiskIoMetrics {
    pub read_bytes_per_sec: u64,
    pub write_bytes_per_sec: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TcpConnectionMetrics {
    pub established: u32,
    pub time_wait: u32,
    pub listen: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GpuMetrics {
    pub utilization_percent: Option<f64>,
    pub vram_used_bytes: Option<u64>,
    pub vram_total_bytes: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessInfo {
    pub pid: u32,
    pub user: String,
    pub cpu_percent: f64,
    pub mem_percent: f64,
    pub vsz_bytes: u64,
    pub rss_bytes: u64,
    pub command: String,
    pub state: String,
}
